### ISSUE TITLE
tf.distribute.MirroredStrategy for asynchronous training

### ISSUE THREAD

**Issue by christianhacker** [tensorflow/tensorflow#57228]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Feature Request

### Tensorflow Version

2.8.1

### Python version

3.8.13

### CUDA/cuDNN version

11.8

### Use Case

I need to run multiple asynchronous copies of the same model on different slices of the dataset (e.g. with bootstrap sampling). There's no *good* way to do this in keras api that I'm aware of, although a couple of hacks exist. Would this use case be feasible with tf.distribute?

### Feature Request

`tf.distribute.MirroredStrategy` is a synchronous, data parallel strategy for distributed training across multiple devices on a single host worker.

Would it be possible to modify this strategy to allow for asynchronous training of all model replicas, without computing the average gradient over all replicas to update weights? In this case each replica would need its own un-mirrored copy of model weights, and the update rule would depend only on the loss and gradients of each replica.

Thanks

**Comment by tilakrayal** [comment]
@christianhacker,
_**`tf.distribute.experimental.ParameterServerStrategy`**_ as an asynchronous CPU/GPU multi-worker solution, where the parameters are stored on parameter servers, and workers update the gradients to parameter servers asynchronously.

Could you please take a look at this doc [link1](go/tf-multi-worker#current-status-of-cpugpu-multi-worker-strategies-with-keras) and [link2](https://www.tensorflow.org/api_docs/python/tf/distribute/experimental/ParameterServerStrategy) and let us know whether you are looking for this feature. Thank you!

**Comment by christianhacker** [comment]
@tilakrayal ,
Thank you for the reply. I've been working on getting ParameterServerStrategy working, but am at an impasse. 

Since this is bootstrap training, I'd like to have access to each worker's model weights post training; I need to track how parameters are different for each bootstrap split (i.e. a random subsample of the original training data with replacement). 

This means that each split should be trained on a separate model replica with the **same initial weights**. Using the chief/worker/param_server model, I do not know whether this means I need a separate worker for *each* split, or whether I can simply define a per-worker dataset function that ensures each per-worker dataset only trains on a single split. 

It's important that the model replicas for each bootstrap split are asynchronous; I need to analyze how the weights change post-training across all splits. It seems that the only way to do this currently is to fetch per-worker model values using a cluster coordinator, possibly with a custom training loop or a callback (if using keras api).

I think the per-worker `get_weights_fn` would need to return a RemoteValue object, then do something with PerWorkerValues, but I am not sure how to implement it. 

Thanks,

**Comment by jamesmullenbach** [comment]
Hi Christian,

It sounds to me like you want to train two models, starting from the same weights, on different datasets (that are subsets of the same dataset), and then compare their weights after training. Maybe a naive question, but is this something you can accomplish by simply training these models in two separate jobs? Is the weight comparison used to change how the models will be trained *during* training?

If the answer to my naive question is no, then this type of use case could be supported by ParameterServerStrategy, with some customization of ClusterCoordinator. As a sketch: we would schedule closures for each worker using the tagged queues, so that each worker gets its own function to run as a training step, which modifies its own specific model. There are some implications of using the tagged queues, though; for instance for `join` to work as expected we'd have to modify it to wait on these tagged closures to finish (which might cause a small performance hit)

**Comment by christianhacker** [comment]
@jamesmullenbach 

You're right; I could just train the models on different jobs. It'd be nice to do so in a coordinated way, since I'm trying to run hundreds of models asynchronously, distributed over at least 10 gpu processes running on a single card. Autosharding the data would be nice too, so I thought `distribute` would be a good fit.

I was able to get something working with Ray; there's some extra overhead but it'll do. Closing unless anyone wants to discuss further.

**Comment by walline** [comment]
@christianhacker 

I'm looking to do something similar to what you were trying. I want to train models asynchronously in parallel with some sparse communication between the models after some timesteps. A feature similar to the MirroredStrategy without the "mirroring" and aggregation of the gradients would probably be what I'm looking for. Did you ever find a solution?

**Comment by christianhacker** [comment]
@walline 

Yes, I wrote a script to automate launching training jobs with the ray framework. It launches several worker processes and handles resource allocation, callbacks, etc., works pretty well.

Not sure about sparse communication between models; I only needed something that could run multiple models in parallel on different splits in the data. It's basically parallel k-folds CV; none of the models are "interacting" to update weights or anything. I'm not sure how to do that with ray.
